source/declarations.tex

```diff
@@ -2869,159 +2869,123 @@
 \indextext{declarator!array}
 
 \pnum
-In a declaration
-\tcode{T}
-\tcode{D}
-where
-\tcode{D}
-has the form
+In a declaration \tcode{T} \tcode{D} where \tcode{D} has the form
 
 \begin{ncsimplebnf}
 \terminal{D1 [} \opt{constant-expression} \terminal{]} \opt{attribute-specifier-seq}
 \end{ncsimplebnf}
 
-and the type of the identifier in the declaration
-\tcode{T}
-\tcode{D1}
-is
-``\placeholder{derived-declarator-type-list}
-\tcode{T}'',
-then the type of the identifier of
-\tcode{D}
-is an array type; if the type of the identifier of \tcode{D}
-contains the \tcode{auto} \grammarterm{type-specifier},
-the program is ill-formed.
-\tcode{T}
-is called the array
-\term{element type};
-this type shall not be a reference type, \cv{}~\tcode{void},
-or a function type.
+and the type of the contained \grammarterm{declarator-id}
+in the declaration \tcode{T} \tcode{D1}
+is ``\placeholder{derived-declarator-type-list} \tcode{T}'',
+the type of the \grammarterm{declarator-id} in \tcode{D} is
+``\placeholder{derived-declarator-type-list} array of \tcode{N} \tcode{T}'',
+where \tcode{N} is the value of
+the \grammarterm{constant-expression}\iref{expr.const};
+if the \grammarterm{constant-expression} is absent,
+the type of the \grammarterm{declarator-id} in \tcode{D} is
+``\placeholder{derived-declarator-type-list} array of unknown bound of \tcode{T}'', except as specified below.
+In either case,
+the type shall not contain the \tcode{auto} \grammarterm{type-specifier}.
+The optional \grammarterm{attribute-specifier-seq} appertains to the array.
+
+\pnum
 \indextext{declaration!array}%
-If the
-\grammarterm{constant-expression}\iref{expr.const}
-is present, it shall be a converted constant
-expression of type \tcode{std::size_t} and
-its value shall be greater than zero.
-The constant expression specifies the
-\indextext{array!bound}%
+A type ``array of \tcode{N} \tcode{T}'' or
+``array of unknown bound of \tcode{T}''
+is an \defn{array type}
+and \tcode{T} is called the array \defn{element type};
+this type shall not be a reference type, \cv{}~\tcode{void}, or a function type.
+\begin{note}
+An array can be constructed from one of the fundamental types
+(except \tcode{void}),
+from a pointer,
+from a pointer to member,
+from a class,
+from an enumeration type,
+or from another array.
+\end{note}
+\begin{example}
+\begin{codeblock}
+float fa[17], *afp[17];
+\end{codeblock}
+declares an array of \tcode{float} numbers and
+an array of pointers to \tcode{float} numbers.
+\end{example}
+
+\pnum
+If the \grammarterm{constant-expression} is present,
+it shall be a converted constant expression of type \tcode{std::size_t}.
```

Comment:
The relative sequencing of this and the above rule forming the type seem to suggest that the conversion to `size_t` is not part of type formation. Can we move this paragraph earlier?

Reply:
I'm splitting up the grammar parsing, otherwise I can't factor the constant-expression stuff properly into the sentence.